### ISSUE TITLE
Fetching entries randomly from Database

### ISSUE THREAD

**Issue by ubbeK** [Starcounter/Starcounter#3761]
Is there a way to get random entries from the Database?

For instance, if I have 1000 entries in the `MyEntry` Database and want to get exactly 10 random entries.
```
    [Database]
    public class MyEntry
    {
        public string Name { get; set; }
    }
```

It does not seem to be efficient to get all entries through Db.SQL request
`var allEntries = Db.SQL<MyEntry>("SELECT i FROM SomeNamepace.MyEntry i);` and then get results randomly from the query result `allEntries`.

cc @miyconst @k-rus

**Comment by miyconst** [comment]
@k-rus do we have something for it? This is a very common functionality for DBMS.

MS-SQL Server supports

```sql
ORDER BY NEWID()
```

MySql Server supports

```sql
ORDER BY RAND()
```

**Comment by k-rus** [comment]
No, we don't have such functionality. Can you describe motivation for the functionality?


> It does not seem to be efficient to get all entries through Db.SQL request
> var allEntries = Db.SQL<MyEntry>("SELECT i FROM SomeNamepace.MyEntry i); and then get results randomly from the query result allEntries.

This is the only way to get correct distribution. Theoretically you can get targeted distribution if you have a unique index with known constant difference between existing key values, but this is not realistic to have. If you don't care about the distribution, then you can use index on, e.g., integer to obtain max and min values and then randomly chose values in the range and use the same index to get values. It will be more efficient than scanning, but it requires a suitable index and no need to control distribution.

**Comment by miyconst** [comment]
There are many cases where such thing can be needed, for example show random ads on a page, or random reviews per product item.

**Comment by k-rus** [comment]
> There are many cases where such thing can be needed, for example show random ads on a page, or random reviews per product item.

You can show the first ad in the index or the first review in the index. Such approach satisfy the requirements.

**Comment by miyconst** [comment]
>You can show the first ad in the index or the first review in the index. Such approach satisfy the requirements.

No, I can't. Because the shown items should be random on each page reload.

**Comment by k-rus** [comment]
> You can show the first ad in the index or the first review in the index. Such approach satisfy the requirements.
> No, I can't. Because the shown items should be random on each page reload.

If it gets the same page it does not mean that it is not random. For example, if normal distribution is chosen for random generation with expected mean value, but the database contains only one value close to the mean value, while the most of the values in tails, then the same page will be shown the most of the time.

I guess you know much more about the requirements, but they are not described.

**Comment by k-rus** [comment]
We discussed this functionality briefly. The actual requirements can be narrowed for each application. For the current application the requirement is to chose show different objects for the same user on every page request. It is not necessary to be always different, but usual. The list of object to chose from varies over time in both directions, i.e., more objects appear in the least or less objects appear in the list.

IMHO, obtaining min and max of some index and then accessing random value in between is a suitable solution.
